### DISCUSSION TITLE
Assigning and updating material property to an element

### DISCUSSION THREAD

**Discussion by sinaomrani96** [idaholab/moose#23903] [Q&A General]
Dear all,
Hope you are doing well.

I wanted to see if we can assign a property to a block, and it gets updated in each element. To make it clear, consider a two region system which in the right-hand side we have a solution which its property (such as density, viscosity, etc.) is dependent on the gas concentration which is injected. I saw a similar question #20334 but I wanted to check if I need to write the material property myself or it has been added.

Best regards,
Sina

**Comment by GiudGiud** [comment]
Hello

This is actually the default for material properties, they are computed for every quadrature point even. 

Constant or constant per-block are optimizations to compute the material properties less-often. 

If your material property definition is simple enough to write in an anaytical expression, you could consider using a ParsedMaterial to try it out, with any local variable dependencies. And then you can use the `output` parameter of the Material to look at its variation in the exodus file

Guillaume

**Reply by sinaomrani96** [reply]
Dear @GiudGiud 

Thanks for your reply. I checked the ParsedMaterial as you said and that's exactly what I needed. My question is that how I can pass this calculated material as a function? I get the property that I need for example density, now I need to pass it to my fluid property module which accepts functions. Or better to ask how I can use this property and not only see it's change in the output. Sorry if it is a basic question.

Thanks.
Sina

**Reply by GiudGiud** [reply]
Which FluidProperty object are you trying to use?

Functions are not supposed to depend on variables, only on time and space. 
However, if the fluid property object queries the functions with x = temperature for example, this can work.
It's what this does:
https://mooseframework.inl.gov/source/userobjects/TemperaturePressureFunctionFluidProperties.html

**Reply by sinaomrani96** [reply]
I am using TemperaturePressureFunctionFluidProperties as you mentioned. It accepts functions as input, so I was wondering if I can pass the property calculated by ParsedMaterial which is dependent on the variable to my fluid property. If this cannot be done, do you have any other suggestion?

**Reply by GiudGiud** [reply]
Oh then you should use ParsedFunctions. 

Please be careful when writing down the expressions
x = temperature
y = Pressure

anything else wont be correct

**Reply by sinaomrani96** [reply]
Thank you for your quick reply dear @GiudGiud. Just to double check then, there is no way to pass a variable-dependent material to the fluid properties to update density, vicosity, and etc? and I need to write my own material and fluid property? Thank you in advance for your time.

**Reply by GiudGiud** [reply]
Nope, no way to pass a material property to a fluid property right now.

It would not be hard to do if you want to add it. 

You do not need to write your own material & fluid property if you are willing to use
TemperaturePressureFunctionFluidProperties + ParsedFunctions
or any of the other pre-made FluidProperties

**Reply by sinaomrani96** [reply]
Unfortunatly, TemperaturePressureFunctionFluidProperties + ParsedFunctions or other FluidProperties do not work for me or I could not find a sutable one. I need to only write a FluidProperty which accepts the ParsedMaterial. I appreciate if you have any advice or similar work suggestion to look at. Thanks.

**Reply by GiudGiud** [reply]
Can you write down your fluid properties here as analytical expressions? I can try to write the ParsedFunctions for you

**Reply by sinaomrani96** [reply]
Oh thank you. I do not have special analytical expression in my mind just wanted to check to see if it works and apply it later.

Here is what I did. I know that you mentioned TemperaturePressureFunctionFluidProperties only works for temperature and pressure correctly but I passed the saturation and it almost gives what I want. However, instead of changing the property at the element that saturation reaches, it changes the whole block. The ParsedMaterial that you mentioned do it right.

Here is my input file.
https://github.com/sinaomrani96/Material-property/blob/main/input.i

Thanks.

**Reply by GiudGiud** [reply]
ok so the constant are fine as they are, you do not need to specify the symbol_names/vars.

If you have pressure OR temperature dependence, as I mentioned you can use x and y in the function expression. 

However, for the pore pressure there's no solution with ParsedFunctions

Creating a new fluid properties that retrieves other material properties will be the way to go if your other objects expert a FluidProperties input

**Reply by sinaomrani96** [reply]
Ok then. Thank you so much. I will look into it to see if I find a similar thing to write the code for the fluid properties. Sorry to ask again, but I appreciate if you have any advice or you know similar approaches and examples to look at.

**Reply by GiudGiud** [reply]
I would start from an existing fluid property object (simpleFP??), then try to sub out one property (for example a constant) for a material property.

Once this compiles, test it before moving on to the next property

**Reply by GiudGiud** [reply]
there's many examples of using material properties. Here's one:
https://github.com/idaholab/moose/blob/next/framework/src/kernels/ADMatReaction.C

notice the addParam<MaterialPropertyName>
and the getADMaterialProperty

**Reply by sinaomrani96** [reply]
Thank you so much @GiudGiud.

**Comment by sinaomrani96** [comment]
Dear @GiudGiud,
Hope you are doing well.

I followed your instruction, and when I compile I get the following error:
error: cannot convert 'const MaterialProperty<double>' to 'libMesh::Real' {aka 'double'} in return
  436 |   return _viscosity1;
      |          ^~~~~~~~~~~
      |          |
      |          const MaterialProperty<double>

Would you please give me an insight into fixing this?

Thanks.
Sina

**Reply by GiudGiud** [reply]
Try return _viscosity1[_qp]
this retrieves the material property at the quadrature point 

also an issue here, the material property must be a reference, not a value 

so in the header you need
Const MaterialProperty & _viscosity1

**Reply by sinaomrani96** [reply]
Dear @GiudGiud,

Thanks for your quick reply. I tried to retrieves the material property at the quadrature point as you mentioned and _viscosity1[_qp] gave me the following error:
error: '_qp' was not declared in this scope; did you mean '_cp'?
  436 |   return _viscosity1[_qp];
      |                      ^~~
      |                      _cp

I changed it to _viscosity1[_cp], and it compiles but when I run my simulation it does not converge at the very beginning.
I run it in debug mode and I get:
Assertion `i < _size' failed
Access out of bounds in MooseArray (i: 4194 size: 216)
at /home/sinaomrani/projects/moose/framework/build/header_symlinks/MooseArray.h, line 276

I appreciate it if you have any suggestion. Sorry if this is a basic question.

**Reply by GiudGiud** [reply]
in which routine is this?

Can you share the code?
_cp is not an index as far as I know.

**Comment by sinaomrani96** [comment]
I am using PorousFlow, and created these two based on SimlpeFluidProperties, and substituting viscosity.

https://github.com/sinaomrani96/Material-property

**Reply by GiudGiud** [reply]
yeah this isnt going to work..
The fluid property doesn't know the indexing into material properties. So it can't retrieve them. 

What's the expression for viscosity?

**Reply by sinaomrani96** [reply]
Oh. I do not have any specific expression in mind, just wanted to connect the viscosity to a variable which I need later. So, what is your suggestion so I can proceed with this?

**Reply by GiudGiud** [reply]
is this variable pressure, temperature or pore pressure?

**Reply by sinaomrani96** [reply]
Actually it is gas saturation and pressure.

**Reply by GiudGiud** [reply]
Are these used in porous flow routinely?
I m thinking there will be a better base class to add your new property there (mu(p, sat))

@cpgr for guidance please

as you can see in SimpleFP and in most properties in the fluid properties module, properties are either:
property(p,T) or property(specific energy, specific volume)

**Reply by sinaomrani96** [reply]
Yes I guess. I will wait to see if @cpgr has a suggestion. You are correct. I am trying to find a better base class. Thanks for your help.

**Reply by cpgr** [reply]
It might be because it is too early in the morning, but I'm not sure exactly what you are trying to do? Is viscosity a nonlinear variable in your model? I'm struggling to see how it depends on saturation?

In any case, the FluidProperties module doesn't really know about variables or the mesh, so probably isn't the right place for this.

If you are using porous flow, could you just make your own fluid properties material (inherit from PorousFlowSingleComponentFluid) and swap out the viscosity calculation with your viscosity variable?

So this bit
```
(*_viscosity)[_qp] = _fp.mu_from_p_T(_porepressure[_qp][_phase_num] * _pressure_to_Pascals,
                                         _temperature[_qp] + _t_c2k) /
                         _pressure_to_Pascals / _time_to_seconds;
```

might become simply
```
(*_viscosity)[_qp] = _viscosity1[_qp]
```

while the rest of the properties come from a fluid properties user object like normal?

**Comment by sinaomrani96** [comment]
Dear @cpgr, good morning then :)))
My viscosity is actually dependent on the gas saturation which is a variable and pressure which is also a variable. So, I used a ParsedMaterial to get this and would like to pass it to viscosity.

I tested your suggestion and I face this error that I cannot fix.
```
error: 'getMaterialProperty' was not declared in this scope; did you mean 'ADMaterialProperty'?
   34 |     _viscosity1(getMaterialProperty<Real>("viscosity1"))
      |                 ^~~~~~~~~~~~~~~~~~~
      |                 ADMaterialProperty
```
Should I add or modify something else?

**Reply by GiudGiud** [reply]
`getMaterialProperty<Real>`

**Reply by sinaomrani96** [reply]
Dear @GiudGiud,

`getMaterialProperty<Real>` also did not work but  `_viscosity1(this->template getMaterialProperty<Real>("viscosity1"))` worked and compiled. However, the simulation does not converge from the beginning even with constant viscosity. I think I should learn more about MOOSE, PorousFlow, and also C++ to solve this but any suggestion from you or Chris is most welcomed. Thanks for the help.

**Reply by cpgr** [reply]
You should be able to modify one of the tests in porous flow that use SimpleFluidProperties (as this uses a constant viscosity). If you then swap in your material with the same constant viscosity you should get the same results. If not, I guess something is incorrect in your implementation. 

I'd be happy to take a look if you link to your code (although it is now night-time here so I won't look until tomorrow)

**Comment by sinaomrani96** [comment]
Sure. That would be great. I worked on it and this is the best that I could get. It compiles, and if I pass a constant value to viscosity1 it works. However, if I call the ParsedMaterial even with a constant value the simulation do not converge at the very beginning. Here is link to the material I wrote based on PorousFlowSingleComponentFluid, and my input file.

https://github.com/sinaomrani96/Material-property

Thanks.

**Reply by GiudGiud** [reply]
You have some inconsistencies there
the computeQpProperty relies half on viscosity1, half on the fp.mu_from_p_T

These lines for example
https://github.com/sinaomrani96/Material-property/blob/09dcc35daf4f4e9bcc4203f6cd185c34642ad3b0/MyFluid.C#L78
https://github.com/sinaomrani96/Material-property/blob/09dcc35daf4f4e9bcc4203f6cd185c34642ad3b0/MyFluid.C#L97

**Reply by sinaomrani96** [reply]
Dear @GiudGiud, Thanks. I fixed those but still facing the same issue.

**Reply by GiudGiud** [reply]
Still some problems:
- AD version, you use raw_value, so you lose the derivative
- non AD version, the derivatives are wrong. dmu/dP is not mu / dt, you need to compute a derivative

**Reply by cpgr** [reply]
Yeah, as @GiudGiud said, your derivatives aren't correct. 

Looking at your example input file, you want a model with 2 phases, a liquid phase (just using water properties like normal), and a gas phase where the viscosity is mu(p, S)? Is that right?

**Reply by sinaomrani96** [reply]
Yes I am working on the derivatives to figure it out.

It may sound a weird right now because I want to brake my problem into simpler one and solve it step by step. Considering this, It is a two phase model which has a liquid phase which has water properties on the left block and a liquid which it viscosity is mu(p, S) on the right block. So, when gas is injected from the left, the right fluid has a constant viscosity until gas reaches to the right block. I thought that maybe assigning same phase to both these liquid is the problem but as I checked this is not the reason. Also model works when a constant viscosity is assigned to the right liquid.

**Reply by sinaomrani96** [reply]
Sorry to make it more clear, I want the viscosity of the liquid on the right block to be dependent on the gas saturation.

**Reply by sinaomrani96** [reply]
Dear @GiudGiud and @cpgr,

I just wanted to give an update that the problem is now solved and the issue was the BCs. Since I was using PorousFlowPiecewiseLinearSink, and it uses the mobility I was facing the problem. When I use use_mobility = false the model works. I need more work on the BCs but in general it is fine. Thanks for your help.
